File: gamewindow.cpp
#include "gamewindow.h"
#include "QObject"

#define FOR_EACH_ROBOT_FUNCTIONS(methodcall)           \
    do{                                                \
    for(int i = 0; i < NUMROBOTS; i++){            \
    robotFunctions[i]->methodcall;             \
    }                                              \
    }while(0);

#define FOR_EACH_MOVER(methodcall)                     \
    do{                                                \
    for(int i = 0; i < NUMROBOTS; i++){            \
    mover[i]->methodcall;                      \
    }                                              \
    }while(0);


gamewindow::gamewindow()
{
    for(int i = 0; i < NUMROBOTS; i++)
    {
        robotFunctions[i] = new GameFunctions();
        mover[i] = new Mover();
    }

    for(int i = 0; i < NUMROBOTS; i++){
        mover[i]->setIndex(i);
        robotFunctions[i]->setIndex(i);
    }
}
void gamewindow::updateInfo(vector<robot> robots, vector<robot> enemy, Point2f def, Point2f atk, dataState ball, KNN k)
{
    deecision.setRobots(robots);
    deecision.setEnemy(enemy);
    deecision.setAreas(def,atk);
    deecision.setBall(ball);
    deecision.setKnnInformation(k);
    deecision.updateObjectives();

    vector<robot> ourTeam = deecision.getTeamRobots(), otherTeam = deecision.getEnemyRobots();
    Point2f centroidDef = deecision.getCentroidDef(), centroidAtk = deecision.getCentroidAtk();

    updateRunFunctions(ourTeam, otherTeam, centroidDef, centroidAtk, deecision.getBall());
}
void gamewindow::updateRunFunctions(vector<robot> robots, vector<robot> enemy, Point2f def, Point2f atk, dataState b)
{
    FOR_EACH_ROBOT_FUNCTIONS(setRobots(robots));
    FOR_EACH_ROBOT_FUNCTIONS(setEnemy(enemy));
    FOR_EACH_ROBOT_FUNCTIONS(setAreas(def,atk));
    FOR_EACH_ROBOT_FUNCTIONS(setBall(b));
    FOR_EACH_ROBOT_FUNCTIONS(setStrategy(STRATEGY));
    FOR_EACH_MOVER(setRobots(robots));
    FOR_EACH_MOVER(setAreas(def,atk));
    FOR_EACH_MOVER(setBall(b));

    FOR_EACH_ROBOT_FUNCTIONS(start());

    FOR_EACH_ROBOT_FUNCTIONS(wait());

    mover[0]->setGameFunctions(robotFunctions[0],robotFunctions[1],robotFunctions[2]);
    mover[1]->setGameFunctions(robotFunctions[0],robotFunctions[1],robotFunctions[2]);
    mover[2]->setGameFunctions(robotFunctions[0],robotFunctions[1],robotFunctions[2]);

    FOR_EACH_MOVER(start());
    FOR_EACH_MOVER(wait());
}

int gamewindow::get_strategy()
{
    return STRATEGY;
}

void gamewindow::EnviaVelocidades(Point2f Velocidades[3]){

    for(int index = 0; index < 3; index ++){
        Velocidades[index].x = mover[index]->getLVel();
        Velocidades[index].y = mover[index]->getRVel();
    }
}

File: gamewindow.py
from robot import *
from decision import *
from gamefunctions import *
from mover import *

class gamewindow:
    def __init__(self):
        self.posBall = list() #point2f
        self.posRobot0 = list() #point2f
        self.posRobot1 = list() #point2f
        self.posRobot2 = list() #point2f
        self.deecision = Decision()
        self.robotFunctions = [GameFunctions(),GameFunctions(),GameFunctions()]
        self.mover = [Mover(),Mover(),Mover()]
        for i in range(NUMROBOTS):
            self.mover[i].setIndex(i)
            self.robotFunctions[i].setIndex(i)

    def updateInfo(self, robots:list, enemy:list, def_:Point2f, atk_:Point2f, ball:dataState):
        self.deecision.setRobots(robots)
        self.deecision.setAreas(def_,atk_)
        self.deecision.setBall(ball)
        self.deecision.setStrategy(STRATEGY)
        self.deecision.updateObjectives()
        teste = self.deecision.getTeamRobots()
        self.updateRunFunctions(teste,self.deecision.getCentroidDef(),self.deecision.getCentroidAtk(),self.deecision.getBall())


    def updateRunFunctions(self, robots:list, def_:Point2f , atk_:Point2f , b:dataState):
        for i in range(NUMROBOTS):
            self.robotFunctions[i].setRobots(robots)
            self.robotFunctions[i].setAreas(def_,atk_)
            self.robotFunctions[i].setBall(b)
            self.robotFunctions[i].setStrategy(STRATEGY)
            self.mover[i].setRobots(robots)
            self.mover[i].setAreas(def_,atk_)
            self.mover[i].setBall(b)
    
        #FOR_EACH_ROBOT_FUNCTIONS(start())
        #FOR_EACH_ROBOT_FUNCTIONS(wait())

        self.mover[0].setGameFunctions(self.robotFunctions[0],self.robotFunctions[1],self.robotFunctions[2])
        self.mover[1].setGameFunctions(self.robotFunctions[0],self.robotFunctions[1],self.robotFunctions[2])
        self.mover[2].setGameFunctions(self.robotFunctions[0],self.robotFunctions[1],self.robotFunctions[2])

        #FOR_EACH_MOVER(start())
        #FOR_EACH_MOVER(wait())



    def EnviaVelocidades(self, Velocidades:list):

        for i in range(3):
            Velocidades[i].x = self.mover[i].getLVel()
            Velocidades[i].y = self.mover[i].getRVel()
        

